packages/themes/tasks/build.js

```diff
@@ -8,93 +8,139 @@
-        const fullTokenName = paramCase(key);
-        const [type, grade] = fullTokenName.split(/(\d+)/g).filter(Boolean);
-        const token = grade ? `${type}-${grade}` : type;
-        return createVariable(token, themes[name][key]);
-      }),
-    };
-  });
-
-  // Mixins
-  let mixins = GENERATED_COMMENT + '\n';
-
-  for (const { name, variables } of values) {
-    const mixin = `@mixin ${NAMESPACE}--theme-${name} {
-  ${variables.map(variable => '' + variable).join('\n  ')}
-}`;
-    mixins += '\n' + mixin;
+  // Create maps for each theme:
+  // $carbon--theme--name: (
+  //   token-name: token-value
+  // ) !default !global;
+  // $theme--name: $carbon--theme--name;
+  const themeMaps = Object.keys(themes)
+    .map(name => {
+      const theme = themes[name];
+      let scssMap = `$carbon--theme--${name}: (`;
+      scssMap += '\n';
+
+      for (const key of Object.keys(theme)) {
+        const name = formatTokenName(key);
+        const value = theme[key];
+        scssMap += `  ${name}: ${value},`;
+        scssMap += '\n';
+      }
+
+      scssMap += ') !default !global;';
+      scssMap += '\n';
+      scssMap += `$theme--${name}: $carbon--theme--${name};`;
+      scssMap += '\n';
+
+      return scssMap;
+    })
+    .join('\n');
+
+  // Variables initialized to default theme
+  // $token-name: map-get($default-theme, token-name) !default !global;
+  const defaultThemeMapName = `$carbon--theme--${defaultTheme}`;
+  const initialVariables = Object.keys(tokens)
+    .map(group => {
+      const groupTokens = tokens[group];
+      let variables = '';
+
+      for (const groupToken of groupTokens) {
+        const name = formatTokenName(groupToken);
+        variables += `$${name}: map-get(${defaultThemeMapName}, ${name}) !default !global;\n`;
+      }
+
+      return variables;
+    })
+    .join('\n');
+
+  // Create carbon--theme mixin, takes a theme as input and reassigns all
+  // theme variables using `!global` before resetting at the end of the function
+  // block
+  let themeMixin = `@mixin carbon--theme($theme: ${defaultThemeMapName}) {\n`;
+
+  // Initialize variables in mixin
+  for (const tokenGroup of Object.keys(tokens)) {
+    for (const token of tokens[tokenGroup]) {
+      const name = formatTokenName(token);
+      themeMixin += `  $${name}: map-get($theme, ${name}) !global;\n`;
+    }
   }
 
+  // Content block
+  themeMixin += '\n';
+  themeMixin += '  @content;';
+  themeMixin += '\n';
+
+  // If block for default theme to reset mixin
+  themeMixin += '\n';
+  themeMixin += `  // Reset to default theme after apply in content`;
+  themeMixin += '\n';
+  themeMixin += `  @if $theme != ${defaultThemeMapName} {
+    @include carbon--theme();
+  }`;
+  themeMixin += '\n';
+  themeMixin += '}';
+
+  let themeMixinAlias = `@mixin theme($args...) {
+  @include carbon--theme($args...) {
+    @content;
+  }
+}`;
+
+  // Files
+  const variables = `${FILE_BANNER}
+${themeMaps}
+
+${initialVariables}`;
```

Comment:
Wondering if we can separate this portion to a separate `.scss` file or surround this part by a mix-in, in order to avoid [`@import 'vendor/@carbon/elements/scss/themes/mixins'`](https://github.com/IBM/carbon-components/blob/v9.83.0/src/globals/scss/_theme-tokens.scss#L13) defining theme variables. Thanks!

Reply:
@asudoh unfortunately for `carbon--themes` we need a reference to the default theme (white) in order to properly reset the theme values 😞

Reply:
Actually, i'm not sure. The problem is that `carbon--theme` requires knowledge about the default theme in order to work, which means that themes needs variables already initialized.

Reply:
Good point, I think the concern is about `$carbon--theme--white` reference in `@mixin carbon--theme($theme: $carbon--theme--white)` is it correct? From my PoV `$carbon--theme--*` including `$carbon--theme--white` can be left in what `_mixins.scss` imports. My concern is rather about each theme tokens defined as glonal, e.g. `$ui-01` - Wanted to make sure loading `_mixins.scss` won't automatically define `$ui-01`, etc.